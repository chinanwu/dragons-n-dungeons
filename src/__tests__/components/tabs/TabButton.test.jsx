import { shallow } from 'enzyme';
import React, { useEffect, useRef } from 'react';

import {
  TabButton,
  mapStateToProps,
} from '../../../components/tabs/TabButton.jsx';

jest.unmock('../../../components/tabs/TabButton.jsx');

jest.mock('react', () => ({
  ...jest.requireActual('react'),
  useEffect: jest.fn(),
  useRef: jest.fn(current => ({ current })),
}));

describe('TabButton component', () => {
  describe('rendering', () => {
    it('renders TabButton', () => {
      const wrapper = shallow(<TabButton id="test" />);
      expect(wrapper.getElement()).toMatchSnapshot();
    });

    it('renders disabled TabButton', () => {
      const wrapper = shallow(<TabButton id="test" isDisabled={true} />);
      expect(wrapper.getElement()).toMatchSnapshot();
    });

    it('renders active TabButton', () => {
      const wrapper = shallow(<TabButton id="test" isActive={true} />);
      expect(wrapper.getElement()).toMatchSnapshot();
    });

    it('renders themed TabButton', () => {
      const wrapper = shallow(
        <TabButton id="test" isActive={true} theme="dark" />
      );
      expect(wrapper.getElement()).toMatchSnapshot();
    });
  });

  describe('behaviour', () => {
    it('disables click', () => {
      const preventDefault = jest.fn();
      const onClick = jest.fn();
      const wrapper = shallow(
        <TabButton id="test" isDisabled={true} onClick={onClick} />
      );
      wrapper.find('#test').simulate('click', { preventDefault });
      expect(preventDefault).toHaveBeenCalled();
      expect(onClick).not.toHaveBeenCalled();
    });

    it('sets focus when is active', () => {
      const focus = jest.fn();
      useRef.mockReturnValue({ current: { focus } });
      shallow(<TabButton id="test" isActive={true} />);
      useEffect.mock.calls[0][0]();
      expect(focus).toHaveBeenCalled();
    });

    it('does not set focus when inactive', () => {
      const focus = jest.fn();
      useRef.mockReturnValue({ current: { focus } });
      shallow(<TabButton id="test" isActive={false} />);
      useEffect.mock.calls[0][0]();
      expect(focus).not.toHaveBeenCalled();
    });
  });

  describe('mapStateToProps', () => {
    it('returned mapped properties', () => {
      const theme = 'light';
      const state = { theme };
      expect(mapStateToProps({ theme: state })).toEqual({ theme: theme });
    });
  });
});
