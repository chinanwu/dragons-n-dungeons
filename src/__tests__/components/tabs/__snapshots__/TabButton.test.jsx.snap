// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TabButton component rendering renders TabButton 1`] = `
<div
  aria-controls="tabPanel__undefined"
  aria-selected={false}
  className="TabButton"
  id="test"
  role="tab"
  tabIndex={-1}
/>
`;

exports[`TabButton component rendering renders active TabButton 1`] = `
<div
  aria-controls="tabPanel__undefined"
  aria-selected={true}
  className="TabButton TabButton--active"
  id="test"
  role="tab"
  tabIndex={0}
/>
`;

exports[`TabButton component rendering renders disabled TabButton 1`] = `
<div
  aria-controls="tabPanel__undefined"
  aria-selected={false}
  className="TabButton TabButton--disabled"
  id="test"
  onClick={[Function]}
  role="tab"
  tabIndex={-1}
/>
`;

exports[`TabButton component rendering renders themed TabButton 1`] = `
<div
  aria-controls="tabPanel__undefined"
  aria-selected={true}
  className="TabButton TabButton--active TabButton--dark"
  id="test"
  role="tab"
  tabIndex={0}
/>
`;
